src/v/features/migrators/cloud_storage_config.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "features/migrators/cloud_storage_config.h"
+
+#include "cluster/controller.h"
+#include "cluster/feature_manager.h"
+#include "cluster/topics_frontend.h"
+#include "features/feature_migrator.h"
+#include "features/feature_table.h"
+#include "features/logger.h"
+#include "ssx/future-util.h"
+#include "vlog.h"
+
+#include <seastar/core/sleep.hh>
+
+#include <type_traits>
+
+using namespace std::chrono_literals;
+
+namespace features::migrators {
+
+template<typename T>
+cluster::property_update<T> make_property_set(const T& v) {
+    return cluster::property_update<T>(
+      v, cluster::incremental_update_operation::set);
+}
+
+ss::future<> cloud_storage_config::do_mutate() {
+    cluster::topic_table& topic_table = _controller.get_topics_state().local();
+
+    vlog(featureslog.info, "Checking for topic configs to update...");
+    size_t update_count = 0;
+    for (const auto& i : topic_table.topics_map()) {
+        auto& topic_conf = i.second.metadata.get_configuration();
+
+        auto& props = topic_conf.properties;
+
+        bool si_enabled
+          = (props.shadow_indexing
+             && props.shadow_indexing
+                  != model::shadow_indexing_mode::disabled)
+            || ((!props.shadow_indexing || props.shadow_indexing == model::shadow_indexing_mode::disabled) && config::shard_local_cfg().cloud_storage_enable_remote_write());
+
+        if (
+          props.retention_local_target_bytes.has_value()
+          || props.retention_local_target_ms.has_value()) {
+            // This may be a retry, if we got partway through an upgrade and
+            // were interrupted: do not keep emitting configuration update
+            // events for all topics
+            vlog(
+              featureslog.info,
+              "Tiered storage topic {} has already been updated",
+              i.first.tp);
+            continue;
+        }
+
+        if (si_enabled) {
+            // This topic must have been created prior to the cloud_retention
+            // feature, because we block topic create/alter while this feature
+            // is not yet active.
```

Comment:
I think that at the moment we only block topic alteration while the feature is not yet enabled.

Reply:
Looks like the check in topics_frontend.cc is an `is_active` call?

Reply:
That's right, but the check is only done when updating topic configuration (code [here](https://github.com/redpanda-data/redpanda/blob/2738403e6c3c60e3708d3ea1dc4abed59f70e202/src/v/cluster/topics_frontend.cc#L174)). Topic creations can still go through while feature is not active.

Reply:
The behaviour for new topics is not deterministic (e.g. for a topic created during migration if a controller leadership transfer occurs, it will get migrated, otherwise it won't). A hacky solution here might be to check the serde version of the `topic_configuration` object (assuming that serde preserves old version). Or we can choose to live with this.

Reply:
Ah, I see -- I wrongly thought we had blocked both creations and alterations.

I'm torn between wanting to have strictly defined behavior (block creations) and a soft landing for users if something goes wrong (permit creations, they can fix up topic properties if something goes weird).  I think the latter is the lesser of evils, & it's defensible that during a mixed version upgrade we don't guarantee which set of "rules" for retention property meanings will apply to created topics.

Reply:
I'm leaning towards allowing creations too. Also, a lot of tests start failing when blocking topic creations and they're not trivial to fix.